### DISCUSSION TITLE
Wait, so multiple definitions of a function are allowed?

### DISCUSSION THREAD

**Discussion by alexpaschoaletto** [zephyrproject-rtos/zephyr#75690] [Q&A] [answered]
# TL:DR

Unless I'm missing something really obvious (apologies beforehand), Zephyr seems to be supporting multiple definitions of a function. The same function is defined three times on different subsystems, but no compilation error is thrown if all subsystems are included together. *But shouldn't that be illegal?*

# Context

I am implementing a kernel feature which involves knowing when a thread has switched out of the CPU in order to execute accordingly. As far as I understand, this functionality is used within the kernel through the `#define` below:

```c
file: zephyr/include/zephyr/tracing.h
/**
 * @brief Called before a thread has been selected to run
 */
#define sys_port_trace_k_thread_switched_out()
```

This define then become the same function in 3 different header files:

```c
//file: zephyr/subsys/tracing/ctf/tracing_ctf.h
//file: zephyr/subsys/tracing/user/tracing_user.h
//file: zephyr/subsys/tracing/sysview/tracing_sysview.h

#define sys_port_trace_k_thread_switched_out() sys_trace_k_thread_switched_out()
void sys_trace_k_thread_switched_out(void);
```

Finally, for each corresponding `.c` file we have a different function declaration:

```c
void sys_trace_k_thread_switched_out(void)        //file: zephyr/subsys/tracing/ctf/ctf_top.c
{
	struct k_thread *thread;
	ctf_bounded_string_t name = { "unknown" };
        ...
}
```
```c
void sys_trace_k_thread_switched_out(void)        //file: zephyr/subsys/tracing/sysview/sysview.c
{
	SEGGER_SYSVIEW_OnTaskStopExec();
}
```
```c
void sys_trace_k_thread_switched_out(void)        //file: zephyr/subsys/tracing/user/tracing_user.c
{
	sys_trace_thread_switched_out_user();
}
```

Which is all good if one decides to just enable *one* of these subsystems, as far as I know. Each one is protected by a kconfig option which is opt-in by default, so in order to use them we have to tweak our `prj.conf`. However, what if we accidentally do this?

```c
CONFIG_TRACING=y
CONFIG_TRACING_USER=y
CONFIG_TRACING_CTF=y
CONFIG_SEGGER_SYSTEMVIEW=y
```

The answer is a successful compilation, instead of the classic "Multiple definitions of" error. **BUT WHY?**

I have checked a lot of files in search of an underlying settings override or something like that, but had no luck. All evidence I found so far suggest that indeed all subsystems coexist and are happily declaring their own versions of the function. Trying to apply the same pattern to my files resulted in the aforementioned error, however, so if anyone got a light on the subject I'd be very happy.
Many thanks in advance.

**Comment by yashi** [comment, marked as answer]
You can't. These are in a `choice`:

https://github.com/zephyrproject-rtos/zephyr/blob/6fc525cc20d4fac44572e5938efe8e30baa2a4fb/subsys/tracing/Kconfig#L26-L68

**Reply by alexpaschoaletto** [reply]
Fair enough.
I looked up straight on the `CMakeLists.txt` (which only included the aforementioned files without any conditions other than their respective *CONFIG_X* options) and saw no warning of overrides being thrown in the compilation phase.  The build system seems to be automatically overriding options of a same `choice` and leaving enabled only the last declared option.

IMHO, I think a little work has to be made here. From what I understand the SEGGER and CTF subsystems are not meant to be used by the kernel in any way; instead they are meant for the application developer to know what's going on. That is totally fine, but each subsystem implementation is effectively monopolizing the kernel tracing functions. A better approach would likely be to have a single function definition and hooks of each subsystem function, like this:

```c
void sys_trace_k_thread_switched_out(void)
{
	#ifdef CONFIG_SEGGER_SYSTEMVIEW 
        segger_trace_k_thread_switched_out();
	#endif
	
	#ifdef CONFIG_TRACING_CTF
	ctf_trace_k_thread_switched_out();
	#endif
	
	#ifdef CONFIG_TRACING_USER
        user_trace_k_thread_switched_out();
	#endif
}
```
It's way more scalable than the current solution.
